### ISSUE TITLE
[prefer-readonly-parameter-types] Suggestion: flag to make this rule behave like prefer-const

### ISSUE THREAD

**Issue by thw0rted** [typescript-eslint/typescript-eslint#3897]
**Repro**

```TS

class Foo() {
  public set baz(val: ReadonlyArray<string>) {
    //... other setter things
    this._baz = val.slice();
  }
}

function updateBaz(arr: string[], foo: Foo): void {
  let v0 = arr[0];
  doStuff(v0);
  // ... do other stuff
  foo.baz = arr;
}
```

**Expected Result**

Don't flag the fact that the `foo` parameter is mutable, because I'm *mutating it on purpose*, but *do* flag the fact that `arr` isn't `ReadonlyArray<string>`, because I'm *not* mutating it.

**Actual Result**

If I enable both `prefer-const` and `prefer-readonly-parameter-types`, the above code will flag the `let` declaration because I don't modify `v0`, and it will flag the `arr` parameter because it's not readonly, but it will also flag `foo` even though I want it to be mutable and do in fact mutate it later in the function.  I could disable the parameter check with an `eslint-disable` comment, but I don't think I can fix the check against `foo` without also removing the check against `arr`, which is no good.

**Additional Info**

As it stands this rule is so opinionated that I don't see a way to use it, short of writing your TS like it was Lisp 😂.  I'd also take a way of decorating individual parameters that says "I do definitely plan on mutating this one".

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.3` |
| `@typescript-eslint/parser`        | `4.28.3` |
| `TypeScript`                       | `4.3.5` |

**Comment by bradzacher** [comment]
`prefer-const` is easy because you can use scope analysis to deterministically and statically figure out if a variable has been ***assigned to***. The answer to the question of "is this variable assigned to?" Is unambiguous and easy to answer.

However "has this parameter been mutated?" is impossible to concretely answer. Some are easy to answer (eg "is the arg ever on the left hand side of an assignment expression?") - but most are impossible

For example:
```ts
function foo(arg: string[]) {
  someOtherFunction(arg);
}
```
Is `arg` mutated here? Impossible to tell without having complete visibility of what the called function does.

What about:
```ts
function foo(arg: MyClass) {
  arg.myMethod();
}
```
Is `arg` mutated here? Again it's impossible to tell without having complete visibility of what the called method does.

> As it stands this rule is so opinionated that I don't see a way to use it

Don't mutate function arguments is how you can use it. You're likely building a lot of code with side-effects.
It's a shift in your programming style to make this work, but people do it and use this rule successfully - mutating becomes the exception. It leads to safer and easier to reason about code in a lot of cases because you know your code has fewer side-effects.

**Comment by thw0rted** [comment]
When I filed this, I think I had a different expectation of what an "immutable Foo" looked like.  `Readonly<Foo>` isn't it -- I mean, maybe it would be for the trivial example I posted, but not for almost any non-trivial class.  It gets a lot more complex, and I don't think I want to invest the time and energy to convince the linter that my type (or more likely, convoluted helper-type) is immutable.

> writing your TS like it was Lisp

is not what I had in mind, so I think this rule isn't for me.  That said, I would love a rule to "prefer ReadonlyArray" (as a function parameter, when mutating methods aren't called) -- is that an option?
